certora/specs/RecoveryConfirmationSignatureValidity.spec

```diff
@@ -0,0 +1,131 @@
+}
+
+// This rule the `multiConfirmRecovery` function can only be called with legitimate signatures.
+// It assumes that the harnessed `checkSignatures` function is implemented correctly and reverts if the signatures are invalid.
+rule multiConfirmRecoveryOnlyWithLegitimateSignatures(env e) {
+    address _wallet;
+    address[] _newOwners;
+    uint256 _newThreshold;
+    uint256 walletNonce = currentContract.nonce(_wallet);
+    SocialRecoveryModule.SignatureData[] signatures;
+    bool _execute;
+    bytes32 recoveryHash = getRecoveryHash(
+        e,
+        _wallet,
+        _newOwners,
+        _newThreshold,
+        walletNonce
+    );
+
+    checkSignatures@withrevert(e, _wallet, recoveryHash, signatures);
+    bool signatureCheckSuccess = !lastReverted;
+
+    multiConfirmRecovery(e, _wallet, _newOwners, _newThreshold, signatures, _execute);
+
+    assert signatureCheckSuccess, "Recovery confirmed with invalid signatures";
+}
+
+// This rule checks that the number of approvals counted by the contract is equal to the number of valid signatures.
+rule approvalsCountShouldEqualTheAmountOfSignatures(env e) {
+    require recoveryConfirmationCount == 0;
+    address _wallet;
+    address[] _newOwners;
+    uint256 _newThreshold;
+    uint256 walletNonce = currentContract.nonce(_wallet);
+    SocialRecoveryModule.SignatureData[] signatures;
+    bool _execute;
+    bytes32 recoveryHash = getRecoveryHash(
+        e,
+        _wallet,
+        _newOwners,
+        _newThreshold,
+        walletNonce
+    );
+
+    multiConfirmRecovery(e, _wallet, _newOwners, _newThreshold, signatures, _execute);
+
+    assert signatures.length == assert_uint256(recoveryConfirmationCount), "More approvals counted than valid signatures";
+}
+
+// This rule checks that only supplied signatures and signers are counted as approvals.
+rule noShadowApprovals(env e) {
+    address _wallet;
+    address[] _newOwners;
+    uint256 _newThreshold;
+    uint256 walletNonce = currentContract.nonce(_wallet);
+    SocialRecoveryModule.SignatureData[] signatures;
+    bool _execute;
+    bytes32 recoveryHash = getRecoveryHash(
+        e,
+        _wallet,
+        _newOwners,
+        _newThreshold,
+        walletNonce
+    );
+    address otherAddress;
+    // We need to correctly initialize pre-state to ensure that the `otherAddress` is not a signer and has not confirmed the recovery.
+    require !currentContract.isGuardian(_wallet, otherAddress) && !currentContract.confirmedHashes[recoveryHash][otherAddress];
+
+    multiConfirmRecovery(e, _wallet, _newOwners, _newThreshold, signatures, _execute);
+
+    assert forall uint256 i. 0 <= i && i < signatures.length => currentContract.confirmedHashes[recoveryHash][signatures[i].signer], "Approvals were not correctly set";
+    assert !currentContract.confirmedHashes[recoveryHash][otherAddress], "Other address should not be able to confirm recovery";
+}
+
+rule duplicateSignaturesRevert(env e) {
```

Comment:
nit: Can you add a description for the rule?

Reply:
Could you please help me with creating a description for this rule? The rule seems self-explanatory, and I'm unsure what additional information I can provide.


I shared my opinion about comments/descriptions here: https://github.com/5afe/CandideWalletContracts/pull/35#discussion_r1636105805